yaml/secret.go

```diff
@@ -221,3 +224,311 @@ func (s *StepSecretSlice) UnmarshalYAML(unmarshal func(interface{}) error) error
 
 	return errors.New("failed to unmarshal StepSecretSlice")
 }
+
+// Validate lints if the steps configuration is valid.
+func (s *SecretSlice) Validate(pipeline []byte) error {
+	invalid, isInvalid := errors.New("invalid secret block found"), false
+
+	// iterate through each step and linting yaml tags
+	for i, secret := range *s {
+		// check required name field
+		if len(secret.Name) == 0 && secret.Origin.Empty() {
+			path, err := yaml.PathString(fmt.Sprintf("$.secrets[%d]", i))
+			if err != nil {
+				return err
+			}
+
+			source, err := path.AnnotateSource(pipeline, true)
+			if err != nil {
+				return err
+			}
+
+			invalid = fmt.Errorf("%w: %s", invalid,
+				fmt.Sprintf("no name provided:\n%s\n ", string(source)))
+			isInvalid = true
+		}
+
+		// validate secret by type
+		switch {
+		case strings.EqualFold(secret.Type, constants.SecretRepo):
+			bad, err := secret.validateRepo(pipeline, i)
+			if bad {
+				invalid = fmt.Errorf("%v: %v", invalid, err)
+				isInvalid = true
+			}
+		case strings.EqualFold(secret.Type, constants.SecretOrg):
+			bad, err := secret.validateOrg(pipeline, i)
+			if bad {
+				invalid = fmt.Errorf("%v: %v", invalid, err)
+				isInvalid = true
+			}
+		case strings.EqualFold(secret.Type, constants.SecretShared):
+			bad, err := secret.validateShared(pipeline, i)
+			if bad {
+				invalid = fmt.Errorf("%v: %v", invalid, err)
+				isInvalid = true
+			}
+		case !secret.Origin.Empty():
+			bad, err := secret.validatePlugin(pipeline, i)
+			if bad {
+				invalid = fmt.Errorf("%v: %v", invalid, err)
+				isInvalid = true
+			}
+		}
+	}
+
+	// check if only default error exists
+	if isInvalid {
+		return invalid
+	}
+
+	return nil
+}
+
+// validateRepo is a helper function to lint secrets of type "repo".
+//
+// this function is used to check the fields of secret with the explicit
+// definition yaml style.
+func (s *Secret) validateRepo(pipeline []byte, i int) (bool, error) {
+	invalid, isInvalid := errors.New("invalid secret"), false
+
+	// check if the engine is not a "native" or "vault"
+	if len(s.Engine) != 0 {
+		if !strings.EqualFold(s.Engine, constants.DriverNative) &&
+			!strings.EqualFold(s.Engine, constants.DriverVault) {
+			path, err := yaml.PathString(fmt.Sprintf("$.secrets[%d].engine", i))
+			if err != nil {
+				return isInvalid, err
+			}
+
+			source, err := path.AnnotateSource(pipeline, true)
+			if err != nil {
+				return isInvalid, err
+			}
+
+			invalid = fmt.Errorf("%w: %s", invalid,
+				fmt.Sprintf("invalid engine value:\n%s\n ", string(source)))
+			isInvalid = true
+		}
+	}
```

Comment:
Could we look at moving this validation check for the `engine` key into the upstream `Validate()` function?

If I'm not mistaken, I see the same chunk of code repeated for each helper validate function:

* [validateRepo()](https://github.com/go-vela/types/blob/41d4f9e3b283bc74d34beb4ec2fa1f33580aafa7/yaml/secret.go#L295-L313)
* [validateOrg()](https://github.com/go-vela/types/blob/41d4f9e3b283bc74d34beb4ec2fa1f33580aafa7/yaml/secret.go#L349-L365)
* [validateShared()](https://github.com/go-vela/types/blob/41d4f9e3b283bc74d34beb4ec2fa1f33580aafa7/yaml/secret.go#L414-L430)

I'm thinking this would help reduce the LOC if we move this chunk into the main `Validate()` call 👍

Reply:
I can take a look. I believe I have it per type because I didn't want to have to worry about it with implicit secrets but implicit secrets will just get `native` by default so it might work.

Reply:
Thanks for the suggestion!!

https://github.com/go-vela/types/pull/126/commits/798d5c5e1875e1a42a93a1be240249c158f5a7cb